cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -535,6 +541,166 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void addSchedule(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId,
+                          @PathParam("app-name") String appName,
+                          @PathParam("app-version") String appVersion,
+                          @PathParam("program-type") String programType,
+                          @PathParam("program-name") String programName,
+                          @PathParam("schedule-name") String scheduleName)
+    throws SchedulerException, BadRequestException, NotFoundException, IOException, AlreadyExistsException {
+    doAddSchedule(request, responder, namespaceId, appName, appVersion, programType, programName, scheduleName);
+  }
+
+  private void doAddSchedule(HttpRequest request, HttpResponder responder, String namespaceId, String appId,
+                             String appVersion, String programType, String program, String scheduleName)
+    throws IOException, BadRequestException, NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationId applicationId = new ApplicationId(namespaceId, appId, appVersion);
+    ProgramId programId = new ProgramId(applicationId, ProgramType.valueOfCategoryName(programType), program);
+    ScheduleSpecification scheduleSpecFromRequest;
+    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8)) {
+      // The schedule spec in the request body does not contain the program information
+      scheduleSpecFromRequest = GSON.fromJson(reader, ScheduleSpecification.class);
+    } catch (IOException e) {
+      LOG.debug("Error reading schedule specification from request body", e);
+      throw new IOException("Error reading request body");
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Request body is invalid json: " + e.getMessage());
+    }
+
+    // If the schedule name is present in the request body, it should match the name in path params
+    if (scheduleSpecFromRequest.getSchedule().getName() != null &&
+      !scheduleName.equals(scheduleSpecFromRequest.getSchedule().getName())) {
+      throw new BadRequestException(
+        String.format(
+          "schedule name in the body of the request (%s) does not match the schedule name in the path parameter (%s)",
+          scheduleSpecFromRequest.getSchedule().getName(), scheduleName));
+    }
+
+    // Add program information to the schedule spec from the request
+    ScheduleSpecification scheduleSpec =
+      new ScheduleSpecification(scheduleSpecFromRequest.getSchedule(),
+                                new ScheduleProgramInfo(programId.getType().getSchedulableType(), program),
+                                scheduleSpecFromRequest.getProperties());
+    lifecycleService.addSchedule(programId, scheduleSpec);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  @POST
```

Comment:
I believe this would be better as a PUT request because schedule updating seems like an idempotent operation. Up for discussion, though.
@poornachandra @anew